pkg/controller/daemon/daemon_controller_test.go

```diff
@@ -868,6 +904,34 @@ func TestDaemonSetPodCreateExpectationsError(t *testing.T) {
 	}
 }
 
+func TestGetPodsForDaemonSetUsesIndexer(t *testing.T) {
+	_, ctx := ktesting.NewTestContext(t)
+	podCount := 5000
+	dsCount := 100
+
+	// Use the wrapper function only for this test to setup pod cache
+	dsc, _, _, _ := newTestControllerWithFakeCache(ctx, podCount, dsCount)
+
+	set := &apps.DaemonSet{ObjectMeta: metav1.ObjectMeta{UID: "owner-42"}}
+
+	// Performance test for the custom indexer
+	startIndex := time.Now()
+	podsByIndex, err := dsc.getPodsFromCache(set)
+	indexDuration := time.Since(startIndex)
+	require.NoError(t, err)
+	t.Logf("ByIndex() fetched %d pods in %v", len(podsByIndex), indexDuration)
+
+	// Test if someone changes ByIndex to List
+	startList := time.Now()
+	// Fetching all pods
+	podsByList, err := dsc.podLister.List(labels.Everything())
+	listDuration := time.Since(startList)
+	require.NoError(t, err)
+	t.Logf("List() fetched %d pods in %v", len(podsByList), listDuration)
+	// List calls from podInformerCache is at least 10 times worse than using an Index
+	assert.Greater(t, listDuration, indexDuration*10, "List() should be much slower than ByIndex()")
```

Comment:
IMO this is not a great way to write UT since this test can be fragile if there are meaningful improvement to `podLister.List` function.

Reply:
From the benchmarking that I did, the difference in performance b/w LIST and Index is in the order of `12X` for `5k` pods to `516X` for `500k` pods ( details on this are in PR Description).

So far, it doesn't look like it will flake. Unless there is a change in underlying architecture, its really hard to close that gap IMO.

However i can change the `10X` part to `1X`.

Reply:
I personally don't like this test at all. Unit tests are not a place to validate performance.
Our test env is often heavily starved and even with 1x, I can imagine that being flaky.

If you want to test performance, create a dedicated benchmark to show performance.

Reply:
I agree in general about having a  benchmark around performance, but my motivation was coming from the point of catching correctness like if in future someone changes it back to LIST, wanted to catch it.
But didn't realize we have starvation in test env. 
For now, I can remove it completely and do something later from CL2 similar to what i was mentioning here - https://github.com/kubernetes/kubernetes/pull/130806#issuecomment-2730373783